node/ai-knowledge-assistant/controllers/app.js

```diff
@@ -0,0 +1,180 @@
+ * @enum {string}
+ */
+const EventType = {
+  MESSAGE: 'MESSAGE',
+  ADDED_TO_SPACE: 'ADDED_TO_SPACE',
+  REMOVED_FROM_SPACE: 'REMOVED_FROM_SPACE',
+  CARD_CLICKED: 'CARD_CLICKED',
+}
+
+/**
+ * Chat application logic.
+ */
+class ChatApp {
+  /**
+   * Instantiates the Chat app.
+   * @param {!Object} event The
+   * [event](https://developers.google.com/workspace/chat/interaction-events)
+   * received from Google Chat.
+   */
+  constructor(event) {
+    this.event = event;
+    this.spaceName = event.space.name;
+    this.userName = event.user.name;
+    this.configCompleteRedirectUrl = event.configCompleteRedirectUrl;
+  }
+
+  /**
+   * Executes the Chat app and returns a
+   * [message](https://developers.google.com/workspace/chat/messages-overview)
+   * as a response.
+   * @return {Promise<import('@googleapis/chat').chat_v1.Schema$Message>} A
+   *     message to post back to the space.
+   */
+  async execute() {
+    switch (this.event.type) {
+      case EventType.ADDED_TO_SPACE:
+      case EventType.MESSAGE:
+        return this.handleAddedToSpaceOrMention();
+      case EventType.REMOVED_FROM_SPACE:
+        return this.handleRemovedFromSpace();
+      case EventType.CARD_CLICKED:
+        return this.handleCardClickEvent();
+      default:
+        return {};
+    }
+  }
+
+  /**
+   * Handles the ADDED_TO_SPACE or MESSAGE event by sending back a welcome message.
+   * It also adds the space to storage, queries all messages currently in the space,
+   * and saves all the messages into storage.
+   * @return {Promise<import('@googleapis/chat').chat_v1.Schema$Message>} A
+   *     welcome text message to post back to the space.
+   */
+  async handleAddedToSpaceOrMention() {
+    if (env.logging) {
+      console.log(JSON.stringify({
+        message: 'Saving message history and subscribing to the space.',
+        spaceName: this.spaceName,
+        userName: this.userName,
+      }));
+    }
+    await FirestoreService.createSpace(this.spaceName);
+
+    try {
+      // List and save the previous messages from the space.
+      const messages = await UserAuthChatService.listUserMessages(
+        this.spaceName, this.userName);
+      await FirestoreService.createOrUpdateMessages(this.spaceName, messages);
+
+      // Create space subscription.
+      await UserAuthEventsService.createSpaceSubscription(
+        this.spaceName, this.userName);
+    } catch (e) {
+      if (e.name === 'InvalidTokenException') {
+        // App doesn't have a refresh token for the user.
+        // Request configuration to obtain OAuth2 tokens.
+        return {
+          actionResponse: {
+            type: 'REQUEST_CONFIG',
+            url: generateAuthUrl(this.userName, this.configCompleteRedirectUrl)
+          }
+        };
+      }
+      // Rethrow unrecognized errors.
+      throw e;
+    }
+
+    // Reply with welcome message.
+    const text = 'Thank you for adding me to this space.'
```

Comment:
Suggested welcome message:

Thank you for adding me to this space. I help answer questions based on past conversation in this space. Go ahead and ask me a question!

Reply:
Done.